fix(client): Fix package path in `generate` output

flatplate opened the pull request:
Previously after running generate with a custom output folder on Windows, code snippet that was shown had backslashes in the import statement rather than forward slashes.

This commit fixes that by replacing the backslashes in the path string with forward slashes if the platform is windows.

Fixes https://github.com/prisma/prisma/issues/7809

I am not sure how one would write a test for this, it being a windows specific problem. The reason for the difference is the [`path.relative`](https://nodejs.org/api/path.html#pathrelativefrom-to) call.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma?pullRequest=13031) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma?pullRequest=13031) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma?pullRequest=13031) it.</sub>

janpio commented:
We probably already have tests that generate on the different platforms, but we probably normalized the paths aways as they are different on the platforms because of the difference between `/` and `\` - which means that your change will not become visible in any of the snapshots.

A way to test that then would be to have different snapshots per platform which are not normalized. I do not know though if we already have that anywhere or if that would be a completely new concept in our tests.

aqrln commented on packages/cli/src/Generate.ts:
```diff
@@ -200,9 +203,13 @@ Please run \`prisma generate\` manually.`
       )
       let hint = ''
       if (prismaClientJSGenerator) {
+        const platform = await getPlatform()
```
I think it probably makes more sense as a local variable inside `replacePathSeperatorsIfNecessary` function, not as a parameter.

Also, you don't really need `getPlatform` from `@prisma/sdk` with all of its heavy machinery for your simple use case, just checking if `os.platform() === 'win32'` should be fine, but no harm in using `getPlatform` either of course.

flatplate committed:
fix(client): Fix package path in generate output

Previously after running generate with a custom output
folder on Windows, code snippet that was shown had
backslashes in the import statement rather than forward
slashes.

This commit fixes that by replacing the backslashes in the
path string with forward slashes if the platform is windows.

flatplate commented:
I looked around a little to understand how the testing on different platforms and the normalization work, but unfortunately couldn't find how I could add a non-normalized test.

If you give me a few pointers on adding such tests, I could try that out but it sounds like a bigger decision than I can make 😄

janpio commented:
You are right, we might not actually have that right now 😞 Hopefully @aqrln or @millsp can point you in the right direction though.

Right now you CI runs are failing because of a dependency update. Rebasing on `main` (or merging from there) should fix it - GH UI should show a nice "Update branch" button now I think that should take care of that.

aqrln commented:
> A way to test that then would be to have different snapshots per platform which are not normalized. I do not know though if we already have that anywhere or if that would be a completely new concept in our tests.

I'm afraid there's no easy way to do this right now, yes. This might require some big and non-trivial changes in our testing setup.

Jolg42 committed:
test(cli): check Prisma Client import path on all OS

Related https://github.com/prisma/prisma/pull/13031

Jolg42 committed:
more asserts

Jolg42 commented:
@aqrln I think someone should use the tests I provided in https://github.com/prisma/prisma/pull/13153 and include them here, so we can see the change

aqrln committed:
tests(cli): simplify assertions (remove unnecessary conditionals)

aqrln commented:
@Jolg42 I removed the conditionals because the import path must be the same on all platforms, PTAL

Jolg42 reviewed (approved):
Looks good! 💯
